src/main/scala/com/webtrends/harness/component/akkahttp/websocket/AkkaHttpWebsocket.scala

```diff
@@ -47,66 +49,128 @@ object AkkaHttpWebsocket {
       case Some(compressOpt) =>
-        Flow[TextMessage].map { tx =>
-          ByteString(tx.getStrictText)
+        Flow[Message].map {
+          case tx: TextMessage =>
+            ByteString(tx.getStrictText)
+          case bm: BinaryMessage =>
+            bm.getStrictData
         }
         .via(compressOpt.flow)
         .map(tx => BinaryMessage(tx))
       case None =>
         Flow[Message]
     }
 
-    Flow[Message]
-      .mapAsync(1) {
-        case tm: TextMessage =>
-          for {
-            input <- tryWrap(inputHandler(authHolder, tm))
-            result <- tryWrap(businessLogic(input))
-          } yield {
-            responseHandler(result)
-          }
-      }
-      .viaMat(compressFlow)(Keep.left)
-      .withAttributes(ActorAttributes.supervisionStrategy(onError))
-      .watchTermination() { (_, done) =>
-        done.map { _ =>
-          close(authHolder)
-        }
-      }
+    val sActor = mat.system.actorOf(callbackActor())
+    val sink =
+      Flow[Message].mapAsync(1) {
+        case tm: TextMessage if tm.isStrict ⇒
+          tryWrap(textToInput(authHolder, TextMessage(tm.getStrictText)))
+        case bm: BinaryMessage if bm.isStrict =>
+          tryWrap(textToInput(authHolder, TextMessage(bm.getStrictData.utf8String)))
+      }.to(Sink.actorRef(sActor, CloseSocket(), {err: Throwable => WSFailure(err)}))
+
+    val source: Source[Message, Unit] =
+      Source.actorRef[Message](completionStrategy, failureStrategy,
+        30, OverflowStrategy.dropHead).mapMaterializedValue { outgoingActor =>
+        sActor ! Connect(outgoingActor, authHolder)
+      } via compressFlow
+
+    Flow.fromSinkAndSourceCoupled(sink, source)
   }
 
-  private def close(authInfo: A): Unit =
+  protected def callbackActor(): Props = Props(new SocketActor())
+
+  private def close(lastInput: Option[I]): Unit =
     if (!closed.getAndSet(true))
-      onClose(authInfo)
+      onClose(authHolder, lastInput)
+
+  private def completionStrategy: PartialFunction[Any, CompletionStrategy] = {
+    case Status.Success(s: CompletionStrategy) => s
+    case Status.Success(_)                     => CompletionStrategy.immediately
+    case Status.Success                        => CompletionStrategy.immediately
+  }
+
+  private def failureStrategy: PartialFunction[Any, Throwable] = {
+    case Status.Failure(cause) => cause
+  }
 
-  private def tryWrap[T](input: Future[T]): Future[T] =
+  private def tryWrap(input: => Future[I]): Future[Any] =
     Try(input).recover({ case err: Throwable =>
-      Future.failed[T](err) }).get
-
-  private def onError: PartialFunction[Throwable, Directive] = {
-    case err: Throwable =>
-      if (errorHandler.isDefinedAt(err)) {
-        errorHandler(err) match {
-          case Stop =>
-            close(authHolder)
-            Stop
-          case status => status // We'll continue processing
+      Future.successful(WSFailure(err))
+    }).get
```

Comment:
We should avoid unprotected gets, even if we are certain it won't cause a problem. I also think its more readable if we just do a Try{} match or try/catch